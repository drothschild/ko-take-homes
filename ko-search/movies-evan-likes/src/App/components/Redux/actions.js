export const SET_SEARCH_FILTER = 'SET_SEARCH_FILTER'
export const SET_DECADE_FILTER = 'SET_DECADE_FILTER'
export const ADD_OMDB_DATA = 'ADD_OMDB_DATA'
export const EXPAND_LINE = 'EXPAND_LINE'
