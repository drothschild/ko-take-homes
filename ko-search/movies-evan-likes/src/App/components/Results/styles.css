.results {
  list-style-type: decimal;
}

.results li {
  cursor: pointer;
}

.results div {
  cursor: default;
}

.results li:nth-child(odd) {
  background-color: #f6f6f5;
  border: #fff 1px solid;
}
