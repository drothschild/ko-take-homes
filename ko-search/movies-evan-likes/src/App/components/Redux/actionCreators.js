import axios from 'axios'
import { SET_SEARCH_FILTER, ADD_OMDB_DATA, SET_DECADE_FILTER, EXPAND_LINE } from './actions'

export function setSearchFilter(searchFilter) {
  return { type: SET_SEARCH_FILTER, payload: searchFilter }
}

export function setDecadeFilter(decadeFilter) {
  return { type: SET_DECADE_FILTER, payload: decadeFilter }
}

function addOMDBData(omdbData) {
  return { type: ADD_OMDB_DATA, payload: omdbData }
}

export function expandLine(expanded) {
  return { type: EXPAND_LINE, payload: expanded }
}

export function getOMDBDetails(title, year, id) {
  const dataWithID = {}
  if (localStorage.getItem([`${id}-imdb`]) && localStorage.getItem([`${id}-rating`])) {
    dataWithID.id = id
    dataWithID.imdbID = localStorage[`${id}-imdb`]
    dataWithID.rating = localStorage[`${id}-rating`]
    return addOMDBData(dataWithID)
  }
  return dispatch => {
    axios
      .get(
        `http://www.omdbapi.com/?apikey=36ff4bcc&t=
        ${encodeURIComponent(title)}&y=${year}&type=movie`
      )
      .then(response => {
        if (response.data.Ratings.filter(rating => rating.Source === 'Rotten Tomatoes')[0]) {
          dataWithID.rating = response.data.Ratings.filter(
            rating => rating.Source === 'Rotten Tomatoes'
          )[0].Value
        } else {
          dataWithID.rating = 'Unrated'
        }
        dataWithID.id = id
        dataWithID.imdbID = response.data.imdbID
        localStorage[`${id}-rating`] = dataWithID.rating
        localStorage[`${id}-imdb`] = response.data.imdbID
        dispatch(addOMDBData(dataWithID))
      })
      .catch(error => {
        // I know console error's are frowned upon, but I think this one makes sense here.
        console.error('axios error', error)
      })
  }
}
