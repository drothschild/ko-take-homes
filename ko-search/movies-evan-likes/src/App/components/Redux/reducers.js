import { combineReducers } from 'redux'
import { SET_SEARCH_FILTER, SET_DECADE_FILTER, ADD_OMDB_DATA, EXPAND_LINE } from './actions'

const searchFilter = (state = '', action) => {
  if (action.type === SET_SEARCH_FILTER) {
    return action.payload
  }
  return state
}

const decadeFilter = (state = '', action) => {
  if (action.type === SET_DECADE_FILTER) {
    return action.payload
  }
  return state
}

const omdbData = (state = {}, action) => {
  if (action.type === ADD_OMDB_DATA) {
    return Object.assign({}, state, { [action.payload.id]: action.payload })
  }
  return state
}

const lineExpandedID = (state = '', action) => {
  if (action.type === EXPAND_LINE) {
    return action.payload
  }
  return state
}
const rootReducer = combineReducers({ searchFilter, decadeFilter, omdbData, lineExpandedID })

export default rootReducer
