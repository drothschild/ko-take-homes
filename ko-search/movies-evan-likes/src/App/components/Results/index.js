import React, { Component } from 'react'
import styles from './styles.css'

class Results extends Component {
  render() {
    return <ul className={styles.results} />
  }
}

export default Results
