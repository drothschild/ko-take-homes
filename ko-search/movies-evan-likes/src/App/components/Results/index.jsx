import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import styles from './styles.css'
import MovieLineItem from './MovieLineItem'

const Results = ({ movieData, decadeFilter, searchFilter }) => {
  const movieList = movieData
    .filter(movie => {
      const decade = parseInt(decadeFilter, 10)
      return (
        (decadeFilter === 'all' ||
          decadeFilter === '' ||
          (movie.year <= decade + 9 && movie.year >= decade)) &&
        (searchFilter.length < 2 ||
          movie.title.toUpperCase().indexOf(searchFilter.toUpperCase()) >= 0)
      )
    })
    .sort((a, b) => (a.title > b.title ? 1 : -1))
    .map(movie => {
      const key = encodeURIComponent(movie.title.toLowerCase() + movie.year)
      return <MovieLineItem movie={movie} key={key} id={key} />
    })
  return (
    <ol className={styles.results}>
      {movieList}
    </ol>
  )
}

Results.propTypes = {
  movieData: PropTypes.arrayOf(PropTypes.object).isRequired,
  decadeFilter: PropTypes.string,
  searchFilter: PropTypes.string
}

Results.defaultProps = {
  decadeFilter: 'all',
  searchFilter: ''
}

const mapStateToProps = state => ({
  decadeFilter: state.decadeFilter,
  searchFilter: state.searchFilter
})

export default connect(mapStateToProps)(Results)
