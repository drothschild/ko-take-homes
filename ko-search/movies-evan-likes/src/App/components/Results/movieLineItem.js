import React, { Component } from 'react'
import styles from './styles.css'

class MovieLineItem extends Component {
  constructor(props) {
    super(props)
    this.state = {
      expanded: false,
      loaded: false
    }
  }
  props = {
    movie: Movie,
    omdbData: object
  }

  render() {
    return <div className={styles.movieLineItem} />
  }
}

export default MovieLineItem
