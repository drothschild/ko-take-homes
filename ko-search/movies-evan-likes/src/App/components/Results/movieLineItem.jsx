import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { getOMDBDetails, expandLine } from '../Redux/actionCreators'

class MovieLineItem extends Component {
  componentDidMount() {
    this.props.getOMDBData()
  }

  render() {
    const { movie } = this.props
    let rating
    let title
    let evanSays
    let expandFunction
    if (this.props.imdbID !== '') {
      title = (
        <a
          href={`https://www.imdb.com/title/${this.props.imdbID}`}
          onClick={event => event.stopPropagation()}
        >
          {movie.title}
        </a>
      )
      rating = this.props.rating
    } else {
      title = movie.title
      rating = 'Unrated'
    }

    if (this.props.lineExpandedID === this.props.id) {
      evanSays = (
        <div>
          {movie.evanSays}
        </div>
      )
      expandFunction = this.props.contractLine
    } else {
      evanSays = null
      expandFunction = this.props.expandLine
    }

    return (
      /* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
      <li onClick={expandFunction}>
        {title} ({movie.year}) {rating}
        {evanSays}
      </li>
    )
  }
}

MovieLineItem.propTypes = {
  movie: PropTypes.objectOf(PropTypes.string).isRequired,
  imdbID: PropTypes.string,
  rating: PropTypes.string,
  lineExpandedID: PropTypes.string,
  id: PropTypes.string.isRequired,
  getOMDBData: PropTypes.func,
  contractLine: PropTypes.func,
  expandLine: PropTypes.func
}

MovieLineItem.defaultProps = {
  imdbID: '',
  rating: 'Unrated',
  lineExpandedID: '',
  getOMDBData: () => {},
  contractLine: () => '',
  expandLine: () => {}
}

const mapStateToProps = (state, ownProps) => {
  const omdbData = state.omdbData[ownProps.id] ? state.omdbData[ownProps.id] : {}
  return {
    imdbID: omdbData.imdbID,
    rating: omdbData.rating,
    lineExpandedID: state.lineExpandedID
  }
}

const mapDispatchToProps = (dispatch, ownProps) => ({
  getOMDBData() {
    dispatch(getOMDBDetails(ownProps.movie.title, ownProps.movie.year, ownProps.id))
  },
  expandLine() {
    dispatch(expandLine(ownProps.id))
  },
  contractLine() {
    dispatch(expandLine(''))
  }
})

export const Unwrapped = MovieLineItem
export default connect(mapStateToProps, mapDispatchToProps)(MovieLineItem)
