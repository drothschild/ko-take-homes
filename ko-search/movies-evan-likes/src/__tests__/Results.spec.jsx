import React from 'react'
import { Provider } from 'react-redux'
import { shallow, render } from 'enzyme'
import Results, { Unwrapped as UnwrappedResults } from '../App/components/Results/index'
import store from '../App/components/Redux/store'
import { setSearchFilter, setDecadeFilter } from '../App/components/Redux/actionCreators'
import movieData from '../movieData.json'
import MovieLineItem from '../App/components/Results/MovieLineItem'

test('Results renders correctly', () => {
  const component = shallow(<UnwrappedResults movieData={movieData} />)
  expect(component).toMatchSnapshot()
})

test('Results should without search items render all the movies', () => {
  const component = shallow(<UnwrappedResults movieData={movieData} />)
  expect(movieData.length).toEqual(component.find(MovieLineItem).length)
})

test('Results should render correct number of movies based on Decade Filter', () => {
  const decade = '1960'
  store.dispatch(setDecadeFilter(decade))
  const component = render(
    <Provider store={store}>
      <Results movieData={movieData} />
    </Provider>
  )
  const movieCount = 1
  expect(component.find('li').length).toEqual(movieCount)
})

test('Search filter shorter than 2 characters should be ignored', () => {
  const term = 'g'
  store.dispatch(setDecadeFilter('all'))
  store.dispatch(setSearchFilter(term))
  const component = render(
    <Provider store={store}>
      <Results movieData={movieData} />
    </Provider>
  )
  expect(component.find('li').length).toEqual(movieData.length)
})

test('Results should render correct number of movies based on Search Filter', () => {
  const term = 'girl'
  store.dispatch(setSearchFilter(term))
  const component = render(
    <Provider store={store}>
      <Results movieData={movieData} />
    </Provider>
  )
  const movieCount = 1
  expect(component.find('li').length).toEqual(movieCount)
})

test('Results should render correct number of movies based on Search Filter and Decade Filter', () => {
  const term = 'the'
  const decade = '1990'
  store.dispatch(setSearchFilter(term))
  store.dispatch(setDecadeFilter(decade))
  const component = render(
    <Provider store={store}>
      <Results movieData={movieData} />
    </Provider>
  )
  const movieCount = 1
  expect(component.find('li').length).toEqual(movieCount)
})
