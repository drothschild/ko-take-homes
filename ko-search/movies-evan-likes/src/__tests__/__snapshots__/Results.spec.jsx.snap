// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results renders correctly 1`] = `
<ol>
  <Connect(MovieLineItem)
    id="2001%3A%20a%20space%20odyssey1968"
    movie={
      Object {
        "evanSays": "The sci-fi film that started it all, I can hardly make this list without including this classic. I recently saw this at the San Francisco Symphony Hall, orchestrated by the symphony and choir, and it was incredible.",
        "title": "2001: A Space Odyssey",
        "year": "1968",
      }
    }
  />
  <Connect(MovieLineItem)
    id="apollo%20131995"
    movie={
      Object {
        "evanSays": "One of my favorite space movies of all time. Makes me both incredibly proud as a member of the human race and incredibly frustrated we haven't done anything in space remotely close to what we did in the 60's & 70's since. NASA is awesome, and this film really feels great to watch. Not to mention it has an excellent musical score.",
        "title": "Apollo 13",
        "year": "1995",
      }
    }
  />
  <Connect(MovieLineItem)
    id="arrival2016"
    movie={
      Object {
        "evanSays": "Loved this movie! Cerebral, engaging science fiction at its best. Do yourself a favor and don't watch any trailers or read anything about it before seeing it. You won't be disappointed!",
        "title": "Arrival",
        "year": "2016",
      }
    }
  />
  <Connect(MovieLineItem)
    id="blade%20runner1982"
    movie={
      Object {
        "evanSays": "Another classic science fiction film, Blade Runner is a must-see. It's thought-provoking and engaging, leaving you questioning whether or not Deckard is a hero or an anti-hero. Take your time with this movie and let it soak in.",
        "title": "Blade Runner",
        "year": "1982",
      }
    }
  />
  <Connect(MovieLineItem)
    id="close%20encounters%20of%20the%20third%20kind1977"
    movie={
      Object {
        "evanSays": "Spielberg classic sci-fi. This was one of the first films that really got me into the genre and it's a classic.",
        "title": "Close Encounters of the Third Kind",
        "year": "1977",
      }
    }
  />
  <Connect(MovieLineItem)
    id="coherence2013"
    movie={
      Object {
        "evanSays": "Another fantastic science fiction, shot on an insanely low budget. Fun fact: there was actually no script for this movie. The director gave each actor/actress note cards the day-of filming with character motivations and suspicions, and the actors improvised the actual dialogue and acting. So, so, SO GOOD!",
        "title": "Coherence",
        "year": "2013",
      }
    }
  />
  <Connect(MovieLineItem)
    id="district%2092009"
    movie={
      Object {
        "evanSays": "Unapologetic and socially charged, District 9 is also smart and visually stunning--particularly when you see the budget spent on the movie. Plus, you gotta love Sharlto Copley's entrance to the silver screen. Love that guy.",
        "title": "District 9",
        "year": "2009",
      }
    }
  />
  <Connect(MovieLineItem)
    id="edge%20of%20tomorrow2014"
    movie={
      Object {
        "evanSays": "",
        "title": "Edge of Tomorrow",
        "year": "2014",
      }
    }
  />
  <Connect(MovieLineItem)
    id="ex%20machina2015"
    movie={
      Object {
        "evanSays": "The animation and acting in this movie are amazing, so is the unsettling nature of the plot. It doesn't shy away from the more difficult questions of what it would mean for man to make AI and what our relationship to it might be.",
        "title": "Ex Machina",
        "year": "2015",
      }
    }
  />
  <Connect(MovieLineItem)
    id="gattaca1997"
    movie={
      Object {
        "evanSays": "With the discovery of CRISPR, I feel like this might not be too far-future science fiction. Very much a social commentary on what it would mean for humans to be genetically designed. Great acting and an engaging plot.",
        "title": "Gattaca",
        "year": "1997",
      }
    }
  />
  <Connect(MovieLineItem)
    id="ghost%20in%20the%20shell1995"
    movie={
      Object {
        "evanSays": "This is one of the best classical science fiction films ever. There's a live-action remake coming out soon that also looks pretty good, but it won't beat the original. The film is as much art as it is entertainment. There's an excellent NerdWriter analysis of this on YouTube that I highly recommend watching.",
        "title": "Ghost in the Shell",
        "year": "1995",
      }
    }
  />
  <Connect(MovieLineItem)
    id="inception2010"
    movie={
      Object {
        "evanSays": "",
        "title": "Inception",
        "year": "2010",
      }
    }
  />
  <Connect(MovieLineItem)
    id="interstellar2014"
    movie={
      Object {
        "evanSays": "I know this movie got a lot of flak when it came out, with many disliking it passionately. I first saw it in IMAX and I was really emotionally moved by it's poignancy. I think the film posits the idea that the universe is so incomprehensibly huge that the only way to make sense of our place in it is with the human relationships we have with each other. The whole \\"love will guide me\\" trope is cliché, agreed, but that isn't going to stop me from really enjoying this moive for the spectacle it offers.",
        "title": "Interstellar",
        "year": "2014",
      }
    }
  />
  <Connect(MovieLineItem)
    id="looper2012"
    movie={
      Object {
        "evanSays": "Another fantastic movie that's a bit more action-y than others on this list. You think you know where it's going, then it throws you for a ringer. Or looper. :P",
        "title": "Looper",
        "year": "2012",
      }
    }
  />
  <Connect(MovieLineItem)
    id="melancholia2011"
    movie={
      Object {
        "evanSays": "Very art-haus, as you would expect from Von Trier. The movie is as close as you can get to the physical manifestation of depression, and it is very depressing to watch. But it is beautiful and haunting.",
        "title": "Melancholia",
        "year": "2011",
      }
    }
  />
  <Connect(MovieLineItem)
    id="moon2009"
    movie={
      Object {
        "evanSays": "Basically a one-man show with Sam Rockwell, this is a fantastic movie and directorial debut by Duncan Jones. You'll enjoy it most if you go in not knowing anything. One of the things I love most about this film is how it plays with your understanding of \\"the twist\\": this movie is smart.",
        "title": "Moon",
        "year": "2009",
      }
    }
  />
  <Connect(MovieLineItem)
    id="patema%20inverted2013"
    movie={
      Object {
        "evanSays": "I actually found out about this movie by seeing a movie poster for a screening on the sidewalk in SF randomly, and I'm so glad I followed through to finding it an watching it. A fantastic (and fantastical) science fiction story that starts slow then really starts to screw with your brain. Gravity is one of those things that we take for granted that it's really cool to see a movie play with our basic understanding of it.",
        "title": "Patema Inverted",
        "year": "2013",
      }
    }
  />
  <Connect(MovieLineItem)
    id="pi1998"
    movie={
      Object {
        "evanSays": "One of the random movies I picked up at Blockbuster (haha, remember those?) back in the day because the cover looked cool. Aaronofsky's tale about the mathematician who discovers the pattern in π, realizes how powerful it is, and subsequently goes mad. It's a rough movie to watch (visually), but its primal nature is part of its appeal.",
        "title": "Pi",
        "year": "1998",
      }
    }
  />
  <Connect(MovieLineItem)
    id="primer2004"
    movie={
      Object {
        "evanSays": "The best time travel movie ever. Full Stop. Nothing else comes close. There's nothing more I need to say about this.",
        "title": "Primer",
        "year": "2004",
      }
    }
  />
  <Connect(MovieLineItem)
    id="solaris2002"
    movie={
      Object {
        "evanSays": "I've also watched the classic version of this too, but for some reason I prefer the remake, personally. It's hauntingly beautiful and the soundtrack is ASMR-inducing. It's a very slow movie (the original is even slower), so be warned.",
        "title": "Solaris",
        "year": "2002",
      }
    }
  />
  <Connect(MovieLineItem)
    id="source%20code2011"
    movie={
      Object {
        "evanSays": "Duncan Jones' second film, equally smart science fiction as Moon. A really fun action-y mental trip through a seemingly normal sequence of events that gets progressively weirder as you dive deeper into the rabbit hole. Great film, great acting, great directing.",
        "title": "Source Code",
        "year": "2011",
      }
    }
  />
  <Connect(MovieLineItem)
    id="the%20abyss1989"
    movie={
      Object {
        "evanSays": "The scene at the end with Ed Harris screaming at Mary Elizabeth Mastrantonio to come back to life gets me every... single... time. Then to follow it up with Harris' deep dive (and the dialog that follows with Mastrantonio)... ugh. What a tear-jerker!",
        "title": "The Abyss",
        "year": "1989",
      }
    }
  />
  <Connect(MovieLineItem)
    id="the%20girl%20who%20leapt%20through%20time2006"
    movie={
      Object {
        "evanSays": "I wasn't expecting to like this movie as much as I did. It feels at first like your average slice-of-life film about high schoolers, but the deeper you get into the movie, the more you realize how interconnected things are and how layered the film has been from the start. Like all great time-travel movies, the depth and intersection of the plot lines keeps you really engaged.",
        "title": "The Girl Who Leapt Through Time",
        "year": "2006",
      }
    }
  />
  <Connect(MovieLineItem)
    id="the%20martian2015"
    movie={
      Object {
        "evanSays": "I absolutely loved the book, and the movie turned out much better than I was expecting it would. If you haven't read the book, I highly recommend it, but the movie stays truthful (for the most part) to the book. Another movie about how awesome NASA is (or could be).",
        "title": "The Martian",
        "year": "2015",
      }
    }
  />
</ol>
`;
